File: CCC21S4.java
import java.io.*;
import java.util.*;
/**
 * CCC '21 S4 - Daily Commute
 * Question type: Graph Theory
 * 7/15 on DMOJ, TLE on Batch 5
 * Question URL: https://dmoj.ca/problem/ccc21s4
 * @author Tommy Pang
 */
public class CCC21S4 {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static PrintWriter pr = new PrintWriter(new BufferedWriter(new OutputStreamWriter(System.out)));
    static StringTokenizer st;
    static int mod = (int) 1e9+7, MM = 200005, N, W, D;
    static Map<Integer, Integer>[] disToDest = new Map[MM];
    static int[] permutation = new int[MM];
    static List<Integer> [] walk = new List[MM];
    static {
        for (int i = 0; i < MM; i++) walk[i] = new ArrayList<>();
        for (int i = 0; i < MM; i++) disToDest[i] = new HashMap<>();
    }
    public static void main(String[] args) throws IOException {
        N = readInt(); W = readInt(); D = readInt();
        for (int i = 0; i < W; i++) walk[readInt()].add(readInt());
        for (int i = 1; i <= N; i++) permutation[i] = readInt();
        for (int i = 1; i <= N ; i++) {
            BFS(i);
        }
        for (int i = 0; i < D; i++) {
            int fir = readInt(), sec = readInt();
            int temp = permutation[fir];
            permutation[fir] = permutation[sec];
            permutation[sec] = temp;
            int ans = Integer.MAX_VALUE;
            for (int j = 1; j <= N; j++) {
                if (permutation[j]==N) {
                    ans = Math.min(ans, j-1);
                }
                else ans = Math.min(ans, j-1+disToDest[permutation[j]].getOrDefault(N, Integer.MAX_VALUE/2));
            }
            System.out.println(ans);
        }

    }
    public static void BFS(int i) {
        Queue<Integer> queue = new LinkedList<>();
        queue.add(i); disToDest[i].put(i, 0);
        while (!queue.isEmpty()) {
            int cur = queue.poll();
            for (int nxt : walk[cur]) {
                if (disToDest[i].get(cur)+1<disToDest[i].getOrDefault(nxt, Integer.MAX_VALUE/2)) {
                    queue.add(nxt);
                    disToDest[i].put(nxt, disToDest[i].get(cur)+1);
                }
            }
        }
    }

    static String next() throws IOException {
        while (st == null || !st.hasMoreTokens())
            st = new StringTokenizer(br.readLine().trim());
        return st.nextToken();
    }
    static long readLong() throws IOException {
        return Long.parseLong(next());
    }
    static int readInt() throws IOException {
        return Integer.parseInt(next());
    }
}

File: CCC21S4.cpp
/**
 * CCC '21 S4 - Daily Commute
 * Question type: Graph Theory
 * 15/15 on DMOJ
 * Question URL: https://dmoj.ca/problem/ccc21s4
 * @author Tommy Pang
 */
#include <bits/stdc++.h>
using namespace std;
const int MM = 2e5+5;
typedef pair<int, int> pi;
int N, W, D, s[MM], idx[MM], dis[MM]; vector<int> adj[MM]; bool vis[MM]; set<pi> st;
void bfs(int st){
    fill(dis, dis+N+1, 1e9);
    queue<int> q; q.push(st); dis[st]=0; vis[st]=true;
    while(!q.empty()){
        int u = q.front(); q.pop();
        for(int v: adj[u]){
            if(!vis[v]) { q.push(v); dis[v] = dis[u] + 1; vis[v] = true;}
        }
    }
}
int main(){
    ios::sync_with_stdio(0); cin.tie(0);
    cin >> N >> W >> D;
    for(int i=1, u, v; i<=W; i++){
        cin >> u >> v; adj[v].push_back(u);
    }
    bfs(N);
    for(int i=1; i<=N; i++){
        cin >> s[i];  idx[s[i]] = i - 1;
    }
    for(int i=1; i<=N; i++){
        st.insert({idx[i] + dis[i], i});
    }
    for(int i=1, x, y; i<=D; i++){
        cin >> x >> y;  int sx = s[x], sy = s[y];
        st.erase({idx[sx]+dis[sx], sx}); st.erase({idx[sy]+dis[sy], sy});
        swap(idx[sx], idx[sy]); swap(s[x], s[y]);
        st.insert({idx[sx]+dis[sx], sx}); st.insert({idx[sy]+dis[sy], sy});
        cout << st.begin()->first << "\n";
    }
}
